File: Haskell/data_structures/01-CheckBrackets.hs

openingBrackets :: [Char]
openingBrackets = ['(', '[', '{']

closingBrackets :: [Char]
closingBrackets = [')', ']', '}']


data OpenedBracket = OpenedBracket { bracket :: Char, position :: Int }


doBracketsMatch :: Char -> Char -> Bool
doBracketsMatch openingBracket closingBracket =
    (openingBracket == '(' && closingBracket == ')') ||
    (openingBracket == '[' && closingBracket == ']') ||
    (openingBracket == '{' && closingBracket == '}')


retrieveFailedOpeningIndexFromStack :: [OpenedBracket] -> Maybe Int
retrieveFailedOpeningIndexFromStack [] = Nothing
retrieveFailedOpeningIndexFromStack ((OpenedBracket _ ix) : _)  = Just ix


findIndexOfNonMatchingBracket :: String -> Maybe Int
findIndexOfNonMatchingBracket string = case failedIndex of Nothing -> retrieveFailedOpeningIndexFromStack remainingStack
                                                           _       -> failedIndex
    where
        go :: [OpenedBracket] -> [(Char, Int)] -> ([OpenedBracket], Maybe Int)
        go stack [] = (stack, Nothing)
        go stack ((letter, ix) : rest)
            | letter `elem` openingBrackets = go ((OpenedBracket letter ix) : stack) rest
            | letter `elem` closingBrackets =
                case stack of [] -> (stack, Just ix)
                              (OpenedBracket openingBracket _) : stackTail ->
                                  if doBracketsMatch openingBracket letter then go stackTail rest
                                  else (stack, Just ix)
            | otherwise = go stack rest
        (remainingStack, failedIndex) = go [] (zip string [1..])


main :: IO()
main = do
    string <- getLine
    case findIndexOfNonMatchingBracket string of Nothing -> print "Success"
                                                 Just ix -> print ix

File: R/data_structures/01_check_brackets.R
#!/usr/bin/env Rscript
openingBrackets <- c("(", "[", "{")
closingBrackets <- c(")", "]", "}")

doBracketsMatch <- function(openingBracket, closingBracket) {
    ((openingBracket == "[" && closingBracket == "]") ||
     (openingBracket == "(" && closingBracket == ")") ||
     (openingBracket == "{" && closingBracket == "}"))
}

retrieveFailedOpeningIndexFromStack <- function(stack, size) {
    if (size == 0) NULL
    else {
        openedBracket <- stack[[size]]
        openedBracket$position
    }
}

findIndexOfNonMatchingBracket <- function(string) {
    stack <- list()
    size <- 0
    for (ix in 1:length(string)) {
        letter <- string[ix]
        if (letter %in% openingBrackets) {
            size <- size + 1
            stack[[size]] <- list(bracket=letter, position=ix)
        } else if (letter %in% closingBrackets) {
            if (size == 0) return(ix)
            openedBracket <- stack[[size]]
            stack[[size]] <- NULL
            size <- size - 1
            if (!doBracketsMatch(openedBracket$bracket, letter)) return(ix)
        }
    }
    retrieveFailedOpeningIndexFromStack(stack, size)
}

Main <- function() {
    conn <- file("stdin", "r")
    string <- readLines(conn, n = 1)
    close(conn)
    charList <- strsplit(string, "")[[1]]
    failedIndex <- findIndexOfNonMatchingBracket(charList)
    cat(if (is.null(failedIndex)) "Success" else as.character(failedIndex), fill = TRUE)
}

if (!interactive()) {
    Main()
}
